raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.raft;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.raft.generated.QuorumStateData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.EOFException;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.util.OptionalInt;
+
+/**
+ * Local file based quorum state store.
+ */
+public class FileBasedStateStore implements QuorumStateStore {
+    private static final Logger LOG = LoggerFactory.getLogger(FileBasedStateStore.class);
+
+    static final int UNDEFINED_EPOCH = -1;
+    private static final int UNKNOWN_LEADER_ID = -1;
+    private static final int NOT_VOTED = -1;
+
+    private final File stateFile;
+
+    private final QuorumStateData cachedState;
+
+    public FileBasedStateStore(final File stateFile) throws IOException {
+        this.stateFile = stateFile;
+        cachedState = loadState(stateFile);
+        LOG.info("Loaded quorum state from local: {}", cachedState);
+    }
+
+    private QuorumStateData loadState(File file) throws IOException {
+        try (final BufferedReader reader = Files.newBufferedReader(file.toPath())) {
+            final ObjectMapper objectMapper = new ObjectMapper();
+            final short version = readShort(reader);
+            final JsonNode jsonNode = objectMapper.readTree(readLine(reader));
+
+            QuorumStateData data = new QuorumStateData();
+            data.fromJson(jsonNode, version);
+            return data;
+        } catch (final NoSuchFileException e) {
+            return new QuorumStateData();
+        } catch (final EOFException e) {
+            delete();
+            LOG.info("Local quorum file corrupted, deleting {}", file.toPath());
```

Comment:
If the quorum state file has become corrupted, do we have a way to recover some of the state from the log? At a minimum, we need to be able to get the latest epoch used in the log, but we should also probably scan the log to locate the latest voter state (this can be done later).